Fix: DeepL (free) fail when more than 2 char lang-code is used, e.g. zh-CN I/O zh

marchage opened the pull request:
~and: Fix for google-free locals that did not translate.~
~Refactored eslint warnings.~

Hi, thank you for the tool. I think I fixed (a) bug(s), at least on my machine it works. ~But that might be a good thing to check out: the paid version of DeepL, _**someone needs to check if it still works**_, I only had the free one at my disposal.~

marchage commented:
Also, perhaps I just got a "weird"/different api end point from DeepL, but that seemed unlikely.  Perhaps just double check.

marchage commented on src/translators/engines/deepl.ts:
```diff
@@ -26,12 +26,26 @@ deepl.interceptors.request.use((req) => {
 
-  req.params = {
-    auth_key: Config.deeplApiKey,
+  // Just add to payload... (not a param, then everything needs to be in params)
```
This might be a remark showing my ignorance on the subject. In my defence: in the http rest extension I was using in vscode I had to do that - put the api key in de request body, in stead of the making it part of the request url as parameter (even though DeepL's example includes both. That actually didn't work for me).

marchage commented on .vscode/deepl-api-test.http:
```diff
@@ -0,0 +1,7 @@
+Content-Length: 54
+Content-Type: application/x-www-form-urlencoded
+
+text=%E6%8B%B7%E8%B4%9D%E8%AE%BE%E7%BD%AE&source_lang=zh&target_lang=en&auth_key={{$dotenv [%]DEEPL_FREE_API_KEY}}
```
(looks okay, but I didn't get it to work in the few minutes I tried. I had to copy the key in :)

antfu replied:
What is this file for?

marchage replied:
"humao.rest-client" VSCode extentension (https://github.com/Huachao/vscode-restclient)

marchage replied:
(perhaps it shouldn't be there...)

marchage replied:
(I wanted to comment, not review, I'll try to get rid of it)

marchage reviewed (commented):
(just a few comments, not really a review I guess)

marchage commented on src/translators/engines/deepl.ts:
```diff
@@ -26,12 +26,26 @@ deepl.interceptors.request.use((req) => {
+      req.data.source_lang = req.data.source_lang.slice(0, 2)
+    // black-listing, not pretty, @TODO white-listing? Something else?
+    const longSupportedLocales = ['EN-GB', 'EN-US', 'PT-PT', 'PT-BR']
+    if (req.data.target_lang && !longSupportedLocales.includes(req.data.target_lang))
```
This guard checks could have been more consistent with the rather elaborate checks in the previous if, hm...

marchage replied:
or the other one could be smaller. perhaps you already guarantee that req.data will contain something.

marchage replied:
This seems wrong, why are they in capitals?! (This has been my first pull request, can you see ;)

KurtGokhan commented:
I was the person who added this API. I confirm that it still works. Maybe your API key or the translated text have some characters that are encoded incorrectly. @marchage, Can you check if that is the case? Maybe there is a zero-width space or something else. I don't see any other reason for why it should not work. Can you test the original request with Postman or something like that.

marchage commented:
I used a VSCode extension "humao.rest-client", https://github.com/Huachao/vscode-restclient, to test it. The sentence it was translating was "Copy settings" in zh-CN I think (let me look that up for you, as well as how it got encoded in my test file as well). But I seem to have forgotten to mention what the crux of this all is! And I put all the changes that I made in this pull request, so that is also sub-optimal (piggy-bagging stuff in). 
Crux: cut the language parameters of at 2 characters like the Google one does. DeepL does not seem to except 5 characters. If you have `nl-BE` or `nl-NL` (in stead of what you likely prescribed somewhere in the manual) `nl`. At least, when I did that, the thing started translating again.
As for the second problem, yeah, I guess I will have to cancel and isolate to individual fixes in stead of piggy bagging stuff on here. I hope I get around to it soon. 
As to Postman, perhaps I can do that later as well...

marchage commented:
api key seemed strange at first: in the url part of their api documentation they put a `>` greater than sign in at the end, but that was wrong/not accepted. But even putting it in params without that only worked when I put everything in params i remember, and the opposite as well: everything in post body. But sure I will try to make it more exact

marchage commented:
(Speaking about piggy-bagging stuff in: I was trying to prevent the extension from wiping my translation files when I set `"i18n-ally.keepFulfilled": true,` and supply it a miss-aligned translation set (like: en-US complete, but hr-HR is missing 106 keys), if I then try and translate those missing keys all my other translation files got emptied of their values. Perhaps it was because I was using DeepL which does not support hr I think?, but it is too complicated because of timing issues/context switches and what not in javascript. So I am not getting a clear stack trace, so never mind for now... perhaps some day :)

marchage commented:
Okay, I reverted that `auth_key` thing I did, that was unneeded (I must have been testing 2 things at once and drawn the wrong conclusion, now it's just about the language codes). I also changed comments, and added some more tests. But I am still piggy-bagging in my Google changes, that might be bad.

I don't have any more time now, back to work.

marchage commented:
Removed other changes for the sake of isolating to what title claims.

marchage commented:
(Why this pull request: I have a big project with source language `zh-CN`. DeepL couldn't handle that as source language parameter. It wanted me to say `zh`)

KurtGokhan commented:
@marchage  I think you are supposed to use this setting:

```
  "i18n-ally.localeCountryMap": {
    "zh-CN": "zh"
  },
```

marchage commented:
Okay, for the source_lang that would be fine. For the target_lang they seem to offer some options: 'EN-GB', 'EN-US', 'PT-PT', 'PT-BR'. But the differences between en-US en en-GB might not be as ...<fill in adjective> as different versions of Chinese writing for example (which they don't support). Yeah, perhaps this setting would have saved me the time, I will try it out later. Tnx.

marchage commented:
But no worries, I've heard enough, I'll close it. Thanks again for looking into it.
